### PR TITLE
Direct export to excel format

### PR DESCRIPTION
**Pull request by stefano**
Implements https://github.com/metabase/metabase/issues/4417

Adds docjure (https://github.com/mjul/docjure) as a dependency to write excel files.

As part of the PR I refactored CSV/JSON exports to reduce code duplication and make it easier to add more export formats in the future.

Question: is the API documentation automatically generated or should I update it manually?

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -78,7 +78,8 @@
                  [ring/ring-json "0.4.0"]                             ; Ring middleware for reading/writing JSON automatically
                  [stencil "0.5.0"]                                    ; Mustache templates for Clojure
                  [toucan "1.0.2"                                      ; Model layer, hydration, and DB utilities
-                  :exclusions [honeysql]]]
+                  :exclusions [honeysql]]
+                 [dk.ative/docjure "1.11.0"]]                         ; Excel export
```
Comment (tlrobinson) [inline project.clj]: Could you put this between `crypto-random` and `environ` so it's alphabetical?

File: src/metabase/api/card.clj [hunk 2]
```diff
@@ -417,26 +417,17 @@
   (binding [cache/*ignore-cached-results* ignore_cache]
     (run-query-for-card card-id, :parameters parameters)))
 
-(defendpoint POST "/:card-id/query/csv"
-  "Run the query associated with a Card, and return its results as CSV. Note that this expects the parameters as serialized JSON in the 'parameters' parameter"
-  [card-id parameters]
+(defendpoint POST "/:card-id/query/:export-format-name"
```
Comment (tlrobinson) [inline src/metabase/api/card.clj]: We could probably shorten `export-format-name` to `export-format`.

**Review by tlrobinson — COMMENTED** [review]
Cool thanks for doing this! I'll let @camsaul review the backend more thoroughly but it's looking good.

File: src/metabase/api/card.clj [hunk 3]
```diff
@@ -417,26 +417,17 @@
...
+  "Run the query associated with a Card, and return its results as a file in the specified format. Note that this expects the parameters as serialized JSON in the 'parameters' parameter"
+  [card-id export-format-name parameters]
   {parameters (s/maybe su/JSONString)}
```
Comment (camsaul) [inline src/metabase/api/card.clj]: I would also add a schema here to validate `export-format`

```clojure
{parameters    (s/maybe su/JSONString)
 export-format (s/enum "csv" "json" "xlsx")}
```

File: src/metabase/api/dataset.clj [hunk 4]
```diff
@@ -46,51 +48,53 @@
...
-    {:status  200
-     :body    (with-out-str
-                ;; turn keywords into strings, otherwise we get colons in our output
-                (csv/write-csv *out* (into [(mapv name columns)] rows)))
-     :headers {"Content-Type" "text/csv; charset=utf-8"
-               "Content-Disposition" (str "attachment; filename=\"query_result_" (u/date->iso-8601) ".csv\"")}}
-    ;; failed query, send error message
-    {:status 500
-     :body   (:error response)}))
+(defn ^:private export-to-csv
```
Comment (camsaul) [inline src/metabase/api/dataset.clj]: Just do `defn-` instead of `defn ^:private`

File: src/metabase/api/embed.clj [hunk 5]
```diff
@@ -274,16 +274,10 @@
...
-  (dataset-api/as-csv (run-query-for-unsigned-token (eu/unsign token) query-params, :constraints nil)))
-
-
-(api/defendpoint GET "/card/:token/query/json"
-  "Like `GET /api/embed/card/query`, but returns the results as JSOn."
-  [token & query-params]
-  (dataset-api/as-json (run-query-for-unsigned-token (eu/unsign token) query-params, :constraints nil)))
+(api/defendpoint GET "/card/:token/query/:export-format-name"
+  "Like `GET /api/embed/card/query`, but returns the results as a file in the specified format."
+  [token export-format-name & query-params]
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: same thing, probably just shorten this to `export-format` and add a schema validation here

File: src/metabase/api/public.clj [hunk 6]
```diff
@@ -75,18 +75,11 @@
   {parameters (s/maybe su/JSONString)}
   (run-query-for-card-with-public-uuid uuid parameters))
 
-(api/defendpoint GET "/card/:uuid/query/json"
-  "Fetch a publically-accessible Card and return query results as JSON. Does not require auth credentials. Public sharing must be enabled."
-  [uuid parameters]
+(api/defendpoint GET "/card/:uuid/query/:export-format-name"
```
Comment (camsaul) [inline src/metabase/api/public.clj]: same thing, probably just shorten this to `export-format` and add a schema validation here

File: src/metabase/models/query_execution.clj [hunk 7]
```diff
@@ -15,11 +15,10 @@
 (def Context
   "Schema for valid values of QueryExecution `:context`."
   (s/enum :ad-hoc
-          :csv-download
+          :download
           :dashboard
           :embedded-dashboard
           :embedded-question
-          :json-download
```
Comment (camsaul) [inline src/metabase/models/query_execution.clj]: @salsakran do we want to stop differentiating between whether a query is downloaded as CSV/JSON/Excel? Seems like it might be useful for usage stats

File: test/metabase/api/card_test.clj [hunk 8]
```diff
@@ -397,8 +397,15 @@
       (perms/grant-native-read-permissions! (perms-group/all-users) database-id)
       ((user->client :rasta) :post 200 (format "card/%d/query/json" (u/get-id card))))))
 
+;;; Tests for GET /api/card/:id/xlsx
+(expect
+  #(> (.length %1) 0)
```
Comment (camsaul) [inline test/metabase/api/card_test.clj]: Just use `count` to get the length of something

```clojure
#(> (count %) 0)
```

**Comment by camsaul** [comment]
I'm not convinced that a test just to make sure the endpoint returns a string with at least one byte is really a good enough test. Can `docjure` be used to parse the file that comes back and transform it to some sort of raw (perhaps CSV) output that we can verify?

**Review by camsaul — CHANGES_REQUESTED** [review]
Code cleanup and better tests

File: src/metabase/models/query_execution.clj (hunk 7 above)
Comment (salsakran) [inline src/metabase/models/query_execution.clj]: strong 👎 
we should add an excel download option.

**Comment by stefano** [comment]
> Can docjure be used to parse the file that comes back and transform it to some sort of raw (perhaps CSV) output that we can verify?

I can use docjure to parse back the excel file and check the cell contents, but the tests http client returns the binary excel file as a string, which corrupts the file (because it converts it to utf8 and bytes that can't be converted get corrupted). I need to see if the http client has an option to return a byte stream or a byte array.

**Comment by stefano** [comment]
@camsaul I've addressed all the code review issues.

**Comment by camsaul** [comment]
Looks good @stefano. There were a couple of linter issues I needed to fix so I opened PR #4922 based on this one so I can make the last few changes. Once CI passes I'll merge that one in.
